File: javascript/QuickSort.js
function partition(items, left, right) {
    //rem that left and right are pointers.

    let pivot = items[Math.floor((right + left) / 2)],
        i = left, //left pointer
        j = right; //right pointer

    while (i <= j) {
        //increment left pointer if the value is less than the pivot
        while (items[i] < pivot) {
            i++;
        }

        //decrement right pointer if the value is more than the pivot
        while (items[j] > pivot) {
            j--;
        }

        //else we swap.
        if (i <= j) {
            [items[i], items[j]] = [items[j], items[i]];
            i++;
            j--;
        }
    }

    //return the left pointer
    return i;
}

function quickSort(items, left, right) {

    if (items.length > 1) {
        let index = partition(items, left, right); //get the left pointer returned

        if (left < index - 1) {
            //more elements on the left side
            quickSort(items, left, index - 1);
        }

        if (index < right) {
            //more elements on the right side
            quickSort(items, index, right);
        }
    }

    return items; //return the sorted array
}

let items = [18, 8, 24, 9, 15, 98, 7, 2, 54, 25, 66, 24, 19, 20];

let output = quickSort(items, 0, items.length - 1);
console.log(output);

File: Python/Sorting/quick_sort.py
def lomuto_partition(arr, p, r):
    pivot = arr[r]
    i = p - 1
    for j in range(p, r):
        if arr[j] <= pivot:
            i += 1
            arr[j], arr[i] = arr[i], arr[j]

    arr[j + 1], arr[i + 1] = arr[i + 1], arr[j + 1]
    return i + 1


def quick_sort(arr, l, r):
    if l < r:
        q = lomuto_partition(arr, l, r)
        quick_sort(arr, l, q - 1)
        quick_sort(arr, q + 1, r)

arr = [2, 8, 7, 1, 3, 5, 6, 4]
quick_sort(arr, 0, len(arr) - 1)
print(arr)
